### PR TITLE
Finetuning Example

### PR DESCRIPTION
**Pull request by nateraw**
## Motivation and Context

Adds:
  - an example for finetuning pretrained torchhub models
  - `MiniKineticsDataModule` from PyTorch Lightning Flash
  - `UCF11DataModule`, a tiny example dataset

Currently a little messy - will fix it up in draft PR and then open it up when I finish.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tutorials/video_classification_example/train.py [hunk 1]
```diff
@@ -430,6 +432,7 @@ def main():
     args = parser.parse_args()
 
     if args.on_cluster:
+        from slurm import copy_and_run_with_config
```
Comment (nateraw) [inline tutorials/video_classification_example/train.py:435]: Thought here is that slurm shouldn't be required for local runs not on cluster.

File: tutorials/video_classification_example/train.py [hunk 2]
```diff
@@ -179,6 +178,9 @@ class KineticsDataModule(pytorch_lightning.LightningDataModule):
     preprocessing transforms and configures the PyTorch DataLoaders.
     """
 
+    TRAIN_PATH = 'train.csv'
```
Comment (nateraw) [inline tutorials/video_classification_example/train.py:181]: Added these attrs so you can override w/ a different file/dir. In my case, I included `MiniKineticsDataModule` that has a slightly different directory structure.

File: tutorials/video_classification_example/train.py [hunk 3]
```diff
@@ -359,7 +361,7 @@ def __getitem__(self, index):
         return next(self.dataset_iter)
 
     def __len__(self):
-        return self.dataset.num_videos()
+        return self.dataset.num_videos
```
Comment (nateraw) [inline tutorials/video_classification_example/train.py:364]: this is a property now, so no need to call

File: tutorials/video_classification_example/train.py [hunk 4]
```diff
@@ -297,11 +299,11 @@ def train_dataloader(self):
         """
         Defines the train DataLoader that the PyTorch Lightning Trainer trains/tests with.
         """
-        sampler = DistributedSampler if self.trainer.use_ddp else RandomSampler
+        sampler = DistributedSampler if (self.trainer is not None and self.trainer.use_ddp) else RandomSampler
```
Comment (nateraw) [inline tutorials/video_classification_example/train.py:302]: this lets you use the datamodule standalone w/ pure pytorch if ya wanted to.

File: tutorials/video_classification_example/finetune.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    def val_dataloader(self):
+        self.val_dataset = self._make_dataset('val')
+        return torch.utils.data.DataLoader(
+            self.val_dataset,
+            batch_size=self.args.batch_size,
+            num_workers=self.args.workers,
+        )
+
+
+class MiniKineticsDataModule(KineticsDataModule):
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: Rather than creating an inheritance hierarchy can you just make these arguments to the original kinetics data module class?

File: tutorials/video_classification_example/finetune.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+import torch
+from pytorchvideo.models.head import create_res_basic_head
+from torch import nn
+from torch.optim import Adam
+
+# HACK
+from train import *
+
+
+class UCF11DataModule(KineticsDataModule):
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: Very strange that a UCF data module would inherit from a Kinetics data module, can you please revise. Favor composition over inheritance.

File: tutorials/video_classification_example/finetune.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+from pytorchvideo.data import LabeledVideoDataset
+from pytorchvideo.data.clip_sampling import UniformClipSampler
+import pytorch_lightning as pl
+import torch
+from pytorchvideo.models.head import create_res_basic_head
+from torch import nn
+from torch.optim import Adam
+
+# HACK
+from train import *
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: Assuming you'll fix this! Feel free to move the kinetics data module out of the other tutorial so that both can use it.

File: tutorials/video_classification_example/finetune.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+
+        # TODO - too many repeated .glob calls here.
+        self.train_paths = []
+        self.val_paths = []
+        self.holdout_scenes = {}
+        for c in self.classes:
+
+            # Scenes within each class directory
+            scene_names = sorted(set(x.name for x in (self.root / c).glob("*") if x.is_dir() and x.name != 'Annotation'))
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: I'm fairly confused about some of these lines, I think it's because i'm not clear on the directory structure you're reading from. What is? E.g. What is the "Annotation" filter

File: tutorials/video_classification_example/finetune.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            
+            # Holdout the last scene
+            # TODO - wrap this in a function so users can override the split logic
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: We don't need users to be able to customize split logic for a tutorial

File: tutorials/video_classification_example/finetune.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+            holdout_scene = scene_names[-1]
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: Why do we only hold out the last scene? Wouldn't it be better to have a 20/80 val/train split?

File: tutorials/video_classification_example/finetune.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        self.frames_per_second = frames_per_second
+        self.clip_duration = (self.num_frames * self.sampling_rate) / self.frames_per_second
+
+        self.classes = [x.name for x in self.root.glob("*") if x.is_dir()]
+        self.id_to_label = dict(zip(range(len(self.classes)), self.classes))
+        self.class_to_label = dict(zip(self.classes, range(len(self.classes))))
+        self.num_classes = len(self.classes)
+
+
+        # TODO - too many repeated .glob calls here.
```
Comment (tullie) [inline tutorials/video_classification_example/finetune.py]: Yeah I agree, we can probably get away with 2 glob calls in this function, one to get the classes, one to get the videos in the class

**Review by tullie — CHANGES_REQUESTED** [review]
Looking forward to getting this tutorial in! Feel free to refactor the previous tutorial slightly to make this one better. E.g. might be worth making a data.py which contains the data module wrappers and common transform recipes.

File: tutorials/video_classification_example/finetune.py (hunk 8 above)
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: Yea this line is definitely messed up haha. Fixed in newer version.

The root dir has directories with different class names. within the class directories there are scene directories that separate out clips shot by the same actor/in the same scene. The `Annotation` dir is not a valid scene, so I'm ignoring it when gathering their names.

ucf11-structure

File: tutorials/video_classification_example/finetune.py (hunk 7 above)
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: yes! already fixed :)

File: tutorials/video_classification_example/finetune.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        self.head = create_res_basic_head(in_features=2048, out_features=self.hparams.num_classes)
+
+        # Metrics
+        self.loss_fn = nn.CrossEntropyLoss()
+        self.train_acc = pl.metrics.Accuracy()
+        self.val_acc = pl.metrics.Accuracy()
+        self.accuracy = {'train': self.train_acc, 'val': self.val_acc}
+
+    def forward(self, x):
+        if isinstance(x, dict):
```
[the remark this answers is not in the document]
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py:54]: Eh, bit of personal preference there. sometimes you just have the video tensor and you don't wanna make a dict just to pass it to the model. thoughts?

File: tutorials/video_classification_example/finetune.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+    def test_step(self, batch, batch_idx):
+        return self.shared_step(batch, "test")
+
+    def configure_optimizers(self):
+        return Adam(self.parameters(), lr=self.hparams.lr)
+
+
+def main():
+    """
```
[the remark this answers is not in the document]
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: of course!

File: tutorials/video_classification_example/finetune.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    parser.add_argument("--audio_mel_step_size", default=16, type=int)
+    parser.add_argument("--audio_num_mels", default=80, type=int)
+    parser.add_argument("--audio_mel_num_subsample", default=128, type=int)
+    parser.add_argument("--audio_logmel_mean", default=-7.03, type=float)
+    parser.add_argument("--audio_logmel_std", default=4.66, type=float)
+
+    # Trainer parameters.
+    parser = pytorch_lightning.Trainer.add_argparse_args(parser)
+    parser.set_defaults(
+        max_epochs=200,
```
[the remark this answers is not in the document]
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py:129]: yup! did it in new version

File: tutorials/video_classification_example/finetune.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+        callbacks=[LearningRateMonitor()],
+        replace_sampler_ddp=False,
+        reload_dataloaders_every_epoch=False,
+    )
+    args = parser.parse_args()
+
+    # Get data, model, configure trainer, and train
+    data = MiniKineticsDataModule(args)
+    model = Classifier(num_classes=6)
```
[the remark this answers is not in the document]
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: yea, in new version I set class attrs in the datamodules and then do `Classifier(num_classes=TheDataModule.num_classes)`

File: tutorials/video_classification_example/finetune.py (hunk 11 above)
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: any help would be appreciated :sweat_smile:

File: tutorials/video_classification_example/train.py (hunk 2 above)
Comment (nateraw) [reply tutorials/video_classification_example/train.py:181]: They aren't always files, though. The mini kinetics datamodule just uses dirs. Perhaps 'train_path_ext' or something?

File: tutorials/video_classification_example/finetune.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tullie) [reply tutorials/video_classification_example/finetune.py:54]: I think either doing a dict or a video tensor is totally fine, for a tutorial we should just choose one though and not have extra conditioning logic. Tutorials should aim for readability over flexibility

File: tutorials/video_classification_example/finetune.py (hunk 10 above)
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: In the [paper](https://www.crcv.ucf.edu/papers/cvpr2009_liu1.pdf), I saw:

>  In the training phase, leave one out cross validation (LOOCV)
scheme is used.

I interpreted that as "pick a random scene from each class and hold it out". Please correct me if I'm wrong!!

File: tutorials/video_classification_example/data.py [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+
+class UCF11DataModule(LabeledVideoDataModule):
+    TRAIN_PATH = None
+    VAL_PATH = None
+    SOURCE_URL = "https://www.crcv.ucf.edu/data/YouTube_DataSet_Annotated.zip"
+    SOURCE_DIR_NAME = "action_youtube_naudio"
+    NUM_CLASSES = 11
+
+    def __init__(self, args):
+        args.verify = False
```
Comment (tullie) [inline tutorials/video_classification_example/data.py:211]: It's a bit hacky setting the args parameter throughout your program, imo it should be a immutable constant that aligns with the args that were sent into the program. Can you add verify as another parameter in LabeledVideoDataModule? Then just pass it in? Also would be fine with it hardcoded to false for all datasets, not too concerned about the ssl certificates tbh.

File: tutorials/video_classification_example/data.py [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+        root = data_path / self.SOURCE_DIR_NAME
+        self.classes = [x.name for x in root.glob("*") if x.is_dir()]
+        self.id_to_label = dict(zip(range(len(self.classes)), self.classes))
+        self.class_to_label = {v: k for k, v in self.id_to_label.items()}
+        self.num_classes = len(self.classes)
+
+        self.train_paths = []
+        self.val_paths = []
+        self.holdout_scenes = {}
+        for c in self.classes:
```
Comment (tullie) [inline tutorials/video_classification_example/data.py:224]: Nice work getting rid of some of the globs, looks good!

File: tutorials/video_classification_example/train.py (hunk 2 above)
Comment (tullie) [reply tutorials/video_classification_example/train.py:181]: Ahh fair enough. Let's just stick with what you have then, TRAIN_PATH is good

File: tutorials/video_classification_example/train.py [hunk 18]
```diff
@@ -81,7 +80,7 @@ def __init__(self, args):
         if self.args.arch == "video_resnet":
             self.model = pytorchvideo.models.resnet.create_resnet(
                 input_channel=3,
-                model_num_class=400,
+                model_num_class=11 # 400,
```
Comment (tullie) [inline tutorials/video_classification_example/train.py:83]: Please fix

File: tutorials/video_classification_example/train.py [hunk 19]
```diff
@@ -233,6 +235,7 @@ def _video_transform(self, mode: str):
             transform=Compose(
                 [
                     UniformTemporalSubsample(args.video_num_subsampled),
+                    Lambda(lambda x: x/255.0),
```
Comment (tullie) [inline tutorials/video_classification_example/train.py:238]:
```suggestion
                    Lambda(lambda x: x / 255.0),
```

**Review by tullie — CHANGES_REQUESTED** [review]
Looking better, please address rest of comments then i'm happy to accept! Thanks for addressing https://github.com/facebookresearch/pytorchvideo/issues/32

File: tutorials/video_classification_example/data.py [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+                data_path=str(
+                    Path(self.root)
+                    / (self.TRAIN_PATH if mode == "train" else self.VAL_PATH)
+                ),
+                clip_sampler=make_clip_sampler(
+                    "random" if mode == "train" else "uniform", self.args.clip_duration
+                ),
+                video_path_prefix=self.args.video_path_prefix,
+                transform=self._make_transforms(mode=mode),
+                video_sampler=DistributedSampler
```
[the remark this answers is not in the document]
Comment (nateraw) [reply tutorials/video_classification_example/data.py:112]: yup! I think my linter did that to me mistakenly

File: tutorials/video_classification_example/data.py [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+        self.holdout_scenes = {}
+        for c in self.classes:
+
+            # Scenes within each class directory
+            scene_names = list(
+                x.name
+                for x in (root / c).glob("*")
+                if x.is_dir() and x.name != "Annotation"
+            )
+            shuffle(scene_names)
```
[the remark this answers is not in the document]
Comment (nateraw) [reply tutorials/video_classification_example/data.py]: The shuffle here is specifically to do random split between train and val - these are scene directory paths **not** video paths directly. Here, we're picking random sub-folders of each class to holdout so we aren't validating on scenes/actors that the model has already seen during training. 

I think it's slightly more clear in the latest version :)

File: tutorials/video_classification_example/data.py (hunk 16 above)
Comment (nateraw) [reply tutorials/video_classification_example/data.py]: Got rid of args

File: tutorials/video_classification_example/finetune.py (hunk 10 above)
Comment (nateraw) [reply tutorials/video_classification_example/finetune.py]: I updated to use 80% of scenes for train, 20% of scenes for val. Note scenes are folders containing videos w/ same scene and/or actor. We dont want to mix videos from same scene folder in train/val splits.

**Review by tullie — APPROVED** [review]
LGTM!

File: tutorials/video_classification_example/models.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+        Args:
+            batch (dict): PyTorchVideo batch dictionary containing a single batch of data.
+            mode (str): The type of step. Can be 'train', 'val', or 'test'.
+
+        Returns:
+            torch.Tensor: The loss for a single batch step.
+        """
+
+        # Pass video tensor through model to get outputs.
```
Comment (tullie) [inline tutorials/video_classification_example/models.py:83]: Can you remove all these inline comments, way to verbose imo

**Comment by tullie** [comment]
@kalyanvasudev can you do a quick pass over this before we merge it in?

**Comment by kalyanvasudev** [comment]
> @kalyanvasudev can you do a quick pass over this before we merge it in?

Sure, on it. Will add my comments by EOD.

File: tutorials/video_classification_example/models.py (hunk 22 above)
Comment (nateraw) [reply tutorials/video_classification_example/models.py]: Done :rocket:

**Comment by nateraw** [comment]
@kalyanvasudev I think we are all good here now - I removed the finetuning tutorial from the site for now + will open a separate PR for it when I finish (likely tomorrow or tuesday :wink: )

**Comment by borijang** [comment]
The finetune script requires command line arguments, can you specify what do we need to input?

**Comment by nateraw** [comment]
Hey there @borijang - I'm not 100% sure this code still works since its so out of date, but the command line args that aren't specified directly with `add_argument` are coming directly from the datamodule. Mainly, the `--root` flag is probably one you'll want to update (but I don't remember since I haven't run this code directly in a while). 

I have similar code [here](https://github.com/nateraw/pytorchvideo-classification) which I think still works that you may find easier to use.
